### DISCUSSION TITLE
Error : "The variable must be elemental" for ForcingFunctionAux AuxKernel - With a digression into Functors

### DISCUSSION THREAD

**Discussion by EdSheltonUKAEA** [idaholab/moose#28627] [Q&A Modules: General]
I've been running into some errors trying to use a combination of Variables, AuxVariables and Materials.  I know not all these things can talk to each other, or use each other as inputs.  

I thought I had found a work around, but my current stumbling block is in an AuxKernel with type ForcingFunctionAux. The error message is "The variable must be elemental".  

I assume this means the AuxVariable I'm using is nodal, but there doesn't seem to be any mention of the elemental variable requirement in the page for the AuxKernel:
https://mooseframework.inl.gov/source/auxkernels/ForcingFunctionAux.html 

Is there a way to define whether an AuxVariable is nodal or elemental?  I can't see anything obvious here:
⁠https://mooseframework.inl.gov/source/variables/AuxVariable.html 

I have created the AuxVariable with default settings.  If it defaults to nodal, is it odd that an AuxKernel wants an elemental variable?

Is there a way to make sure my AuxVariable is elemental for the ForcingFunctionAux, but can still work as a nodal variable where this is needed in for other functions or calculations?

**Comment by EdSheltonUKAEA** [comment]
I've  got an initial answer to this from a colleague.  By setting the AuxVariable to have parameters:
    order = CONSTANT
    family = MONOMIAL
I have made it elemental.  And by changing a few other things I now have a model which is running, but not converging.  This is likely to be because of some other problem.

**Reply by GiudGiud** [reply]
A nodal variable is defined with the dofs at the nodals, for example a lagrange family variable. 
An elemental variable would tentatively be a variable where all the dofs are inside the element.

The pickle is that there are many variable types which have nodal dofs, side dofs and elemental dofs, so the nodal/elemental distinction is not quite enough.

I m not sure why ForcingFunctionAux is elemental-only. I ll take a look

**Reply by GiudGiud** [reply]
so there is nothing "elemental"-only about it, if only that it's been implemented for working on qps
```
  return _u_old[_qp] + _dt * _func.value(_t, _q_point[_qp]);
```
for nodals we would want something like below for the function
```
_func.value(_t, *_current_node);
```

**Reply by EdSheltonUKAEA** [reply]
Thanks for the comments on this.  I'm not sure if the second one means that ForcingFunctionAux should work with Lagrangian AuxVariables.  But even if it "should", for whatever reason in my model if I allow the AuxVariable family to default to LAGRANGE I get the "must be elemental" error when I try to run it.

If I'm stuck with using a MONOMIAL AuxVariable, is there a way of converting this into another AuxVariable which is LAGRANGE and so outputs nodal values?

Or perhaps I can achieve what I want another way.  I'm using ForcingFunctionAux to obtain the value of the AuxVariable at the end of the previous timestep (i.e. start of the current time step) by setting the function, f, to zero, i.e.  V(t+Δt)=V(t)+fΔt = V(t) .

I expect there is a simpler or more elegant way of getting the value at the start of the current timestep of an AuxVariable (or a MaterialProperty might work) at every node.  But looking through the Source Documentation page I haven't found one yet.  I don't think I can use a Postprocessor for this, as I need this value at every node to be available for calculations within the solve.  Postprocessors giving average, integral or extreme values over a domain won't give me what I want.

Is there a way of calling the value at the start of the current time step of an AuxVariable of any type?

**Reply by GiudGiud** [reply]
>  But even if it "should", for whatever reason in my model if I allow the AuxVariable family to default to LAGRANGE I get the "must be elemental" error when I try to run it.

Error is good, it is not currently implemented for nodal variables. I can take a look at making it work for nodal variables if you want. 

> If I'm stuck with using a MONOMIAL AuxVariable, is there a way of converting this into another AuxVariable which is LAGRANGE and so outputs nodal values?

There is the ProjectionAux to go from elemental to nodal variables. This will incur some projection error of course.

> Is there a way of calling the value at the start of the current time step of an AuxVariable of any type?

If you execute (execute_on parameter) the auxkernel that sets that variable on timestep_begin only, it will retrain the value at the start of the time step

**Reply by EdSheltonUKAEA** [reply]
Thank you.  This is really helpful and gives me some other options to try.  So I probably don't need the ForcingFunctionAux in a nodal version right now.

**Reply by EdSheltonUKAEA** [reply]
Using execute_on = timestep_begin looks like it should let me do what I want much more elegantly that using ForcingFunctionAux.  But there doesn't seem to be an execute_on parameter for Material properties.  So ideally I need to convert a material property into a nodal AuxVariable.  However, it looks like this isn't currently possible where is says here: https://mooseframework.inl.gov/source/auxkernels/ParsedAux.html
"Material properties are currently not supported, but it would be really easy to add it so feel free to contact us."
This could be really helpful!  How easy and quick could it be to implement this?  Days? Hours?

Altenatively, I'm looking at the newer(?) concept of Functors, where it looks like Material Properties, AuxVariables and Functions all become more interchangeable as nodal parameters.  What is the first step to either defining a Material property as a Functor or making a copy of a conventional Material property as a Functor?  As I read it, a FunctorMaterial can use as inputs existing FunctorMaterials, but not non-Functors, so I always seem to be one step away from converting my conventional Materials into a Functor then to use with AuxVariables.

**Reply by EdSheltonUKAEA** [reply]
I've just found this: https://mooseframework.inl.gov/source/materials/MaterialFunctorConverter.html
"This Material handles the conversion from functor material properties to regular material properties. **Unfortunately, the conversion in the other direction, from regular material properties to functors, is not implemented** and would be quite challenging notably because of the need to handle complex material property dependency resolution."  [My bold.]
This seems to answer my second question.  So it looks like being able to include MatProps in a ParsedAux is my best option.

**Reply by EdSheltonUKAEA** [reply]
As I read more about Functors and FunctorMaterials, I'm wondering if for the calculations I want to do, using FunctorMaterials might in fact be a better option.  Many of the "Materials properties" I'm using during each time step solve are just temporary parameters used in calculation, not something that needs to be output or plotted, or even saved into the next time step.  As it describes FunctorMaterials as being calculated "on the fly", does this mean using them in calculations like this should be less demanding on memory and make a solve faster?

**Reply by GiudGiud** [reply]
> So ideally I need to convert a material property into a nodal AuxVariable

Keep in mind there are no nodal material properties. You can convert to an elemental variable using a MaterialRealAux

> "Material properties are currently not supported, but it would be really easy to add it so feel free to contact us."
This could be really helpful! How easy and quick could it be to implement this? Days? Hours?

Probably less than a day thing. Seems like the ideal way for me to procrastinate making slides. I'll let you know how it goes.
I m not sure you really need it though? It wont offer that much more than a MaterialRealAux, then chained with a ParsedAux

> Altenatively, I'm looking at the newer(?) concept of Functors

Definitely newer. We have been using them for ~2 years now.

> . What is the first step to either defining a Material property as a Functor or making a copy of a conventional Material property as a Functor?

Regular material properties cannot be defined as functors unfortunately.

> As I read more about Functors and FunctorMaterials, I'm wondering if for the calculations I want to do, using FunctorMaterials might in fact be a better option.

Transitioning from Materials to FunctorMaterials is an option. You just have to take a look at your Materials in the Materials block to see how big of a task it will be to change that code. 

> does this mean using them in calculations like this should be less demanding on memory and make a solve faster?

Pre-computing properties can be an optimization for some problems with expensive constitutive models. So functor materials are not always faster. However, functor materials only compute the 1 property you need, while materials always computed every single properties no matter what.  That can be an improvement

**Reply by GiudGiud** [reply]
https://github.com/idaholab/moose/pull/28673 
please let me know if it does not work. It will error if you try to set a nodal variable, because material properties are not defined at nodes. If you need that, then set an elemental variable first, then use a ProjectionAux to get to a nodal variable

**Reply by EdSheltonUKAEA** [reply]
Sorry for the delay in responding to this.  I was on holiday most of last week.  (I visited Isaac Newton's apple tree!)

I've converted a simplified version of my model to Functor materials properties used the the two-step process of FunctorMaterialRealAux and ProjectionAux to get from elemental FunctorMaterials to nodal AuxVariables and IT WORKS!  Once typos and bugs were sorted out the model ran to completion quickly and smoothly with no time cutbacks.

Now I just need to apply it to my full model with a more substantial set of materials calculations, so there will be quite a lot of editing switching these to Functors, but probably a good exercise to see if parts of it can be simplified or slimmed down anyway.

Thanks for this.  Functors seem to be the answer to my problems.

**Comment by GiudGiud** [comment]
I saw the message in my email. Please keep in mind execute_on for FunctorMaterial relates to refreshing the caching of element and face values. 
It's generally easier to leave that setting to the default, in which case the FunctorMaterial does not use any caching

**Reply by EdSheltonUKAEA** [reply]
Thanks.  I found a fix which by calling the corresponding elemental AuxVariable as the functor to give the value at timestep_begin.
Now I get non-zero values for the rate of change over the timestep, but back to the usual problem of the solve not converging.

The error message is:
>   Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 0
>  Solve Did NOT Converge!
> Aborting as solve did not converge

**Reply by EdSheltonUKAEA** [reply]
Should I use the AuxVariable which duplicates the FunctorMaterial also to give the timestep_end "now" value?

**Reply by GiudGiud** [reply]
Is the value of the auxvariable dependent on the nonlinear variables? Because if so, no derivative information is propagated and the Jacobian would be missing terms. 

Where are you in the process of troubleshooting the nonconvergence? 
wrt this page
https://mooseframework.inl.gov/moose/application_usage/failed_solves.html

**Reply by EdSheltonUKAEA** [reply]
> Is the value of the auxvariable dependent on the nonlinear variables? Because if so, no derivative information is propagated and the Jacobian would be missing terms.

Indirectly, yes.  In my simplified example, x is the non-linear variable being solved.  So the FunctorMaterial is calculated as a function of this.  Then the AuxVariable takes the value of FunctorMaterials using FunctorMaterialRealAux.

> Where are you in the process of troubleshooting the nonconvergence?
> wrt this page
> https://mooseframework.inl.gov/moose/application_usage/failed_solves.html

I'll have to work through that and get back to you!

**Reply by GiudGiud** [reply]
Instead of using an auxiliary variable, could you please use a functor material for this value?
The ADParsedFunctorMaterial is likely flexible enough to replace the auxkernel?
And it will conserve derivatives

**Reply by EdSheltonUKAEA** [reply]
> Instead of using an auxiliary variable, could you please use a functor material for this value?

For which value?  If you mean the value at the start of the timestep, that's what I tried first, but always got zeros for the change over the timestep.  If I understood your previous message correctly, FunctorMaterials "execute_on" don't remember values from the end of the previous timestep, they calculated new each time. 

Should I call back a new FunctorMaterial for the start of the timestep using the AuxVariable at timestep_begian, and then use that is the rate of change calculation?

**Reply by GiudGiud** [reply]
For every value that is using an auxiliary variable right now. 
Since functor materials are evaluated on-the-fly (unless you specified an execute_on, in which case there was caching), the zeros must come from whatever is being input in the functor material. 

If you call a functor material from an auxkernel, you will lose derivative information.
In the troubleshooting instructions, you can look at the Jacobian analyser section, to see if that is the current problem you are facing

**Reply by EdSheltonUKAEA** [reply]
Ultimately, as I need the rate value as a conventional material property to use in ADMatHeatSource kernel, could I call both "then" and "now" values as other conventional materials properties from the AuxKernel and FunctorMaterial respectively, and calculate the difference and rate in the [Materials] block rather than [FunctorMaterials] block?  Would doing it this way provide what's needed for the solve to work?

**Reply by GiudGiud** [reply]
oh you are computing the rate manually?
In a functor material, you can do functor.dot(arg, state) to get the time derivative. 

Computing the rates manually could be done in Materials as well. 
I suspect you would not be able to get second order time integration if you compute rates manually instead of using the dot() / coupledDot routines

**Reply by EdSheltonUKAEA** [reply]
OK.  Thanks.  I'll take a look at these tomorrow as about to head home.

Can you show the "functor.dot(arg, state)" in an example input file?  
But it looks like that might be something in a C++ file, and I don't work with that.  I can only use "off the shelf" MOOSE.

**Reply by GiudGiud** [reply]
functor.dot is indeed some C++

I ll add stuff to the shelf today for you to be able to use that then

**Reply by EdSheltonUKAEA** [reply]
OK.  This is interesting.  
I've used postprocessors to track the values of the following:
- A. ElementIntegralFunctorPostprocessor for FunctorMaterial "fm_x_then", i.e. integrated values of x at start of timestep using ADParsedFunctorMaterial with execute_on = TIMESTEP_BEGIN.
- B. ElementIntegralFunctorPostprocessor for FunctorMaterial "fm_x_now", i.e. integrated values of x at end of timestep using ADParsedFunctorMaterial with execute_on = TIMESTEP_END (or leaving this line out so it uses TIMESTEP_END as the default value, it makes no difference.)
- C. ElementIntegralVariablePostprocessor for elemental AuxVariable "elav_x_now" which takes the value of "fm_x_now" via FunctorMaterialRealAux.
- D. ElementIntegralVariablePostprocessor for nodal AuxVariable "x_now" which takes the value of "elav_x_now" via ProjectionAux.

What I would expect to see is:
- B is different from A because the solve tries new values of the non-linear variable which will give new values of x at the end of the timestep compared with at the beginning.
- B and C are equal because values are just copied to the elemental AuxVariable.
- C and D might be different if there are errors arising from the projection from elemental values to nodal values.

What I get in the postprocessor output is:
- A and B are always exactly equal.
- B and C are always different (if non-zero).  But the differenences are relatively very small, not usually before 4th or 5th significant figure.
- C and D are always exactly equal.

So it looks to me like the execute_on parameter in ElementIntegralFunctorPostprocessor isn't doing anything.  (I also tried using INITIAL instead of TIMESTEP_BEGIN.  Incidentally, does the value of execute_on have to be in captial letters, or is it case-insensitive?)

However, it seems new values of x_now have been calculated as these are transferred to elav_x_now.  Or is the transfer from fm_x_now to elav_x_now more than simply copying values?  Is some calculation done which affects how the integral of the values across the domain is calculated by the postprocessors?

Finally, it seems for the model I'm running to test this, there are no errors from the elemental to nodal projection, at least when integrated over the domain.

What could be going wrong here, or what have I misunderstood?



Edit: Also, if this was all working as I planned, I would expect to see "fm_x_then" equal "fm_x_now" from the previous timestep.  However, it isn't even the case that "fm_x_then" equals "eval_x_now" from the previous timestep.  So although a "new" value seems to have been sent to "eval_x_now" it isn't being used as the "timestep_begin" value for the next timestep by the FunctorMaterial.  Something seems fundamentally wrong with how I am trying to do what should be really simple:
 dx/dt = ( x(t + delta_t) - x(t) ) / delta_t.
It's getting the value of x(t) which appears to be causing the problems.

**Reply by GiudGiud** [reply]
I added this object to let you take the time derivative of a functor in another functor.
https://github.com/idaholab/moose/pull/28756

It is in moose/next right now. I was hoping to get back to you once it gets in moose/master but it may take another day or so to propagate this time. 
You ll need to take a look at the documentation (built locally) or in the source to be mindful of all its limitations. The documentation will not appear online until the object makes it to moose/master. 


> C and D might be different if there are errors arising from the projection from elemental values to nodal values.

The projection minimizing the L2 error, it's actually quite likely the L1 error (roughly what C-D is, but not exactly! since the sign of the elemental-nodal variable might change) is fairly minimal

> B and C are equal because values are just copied to the elemental AuxVariable.

This depends on the timing of execution of the auxkernel as well.

>  execute_on parameter in ElementIntegralFunctorPostprocessor isn't doing anything

It certainly is doing something. You can turn on Debug/show_execution_order = ALWAYS to see this postprocessor execute.
Btw, the execute_on flag of the CSV and Console output also matter here. 

> does the value of execute_on have to be in captial letters, or is it case-insensitive?)

Case-insensitive

> Or is the transfer from fm_x_now to elav_x_now more than simply copying values

Have you paste `fm_x_now` somewhere so I can take a look at what it does and which parameters you passed?
It is likely not copying, rather it forwards an evaluation. FunctorMaterials are not meant to stash values, unless you are using the caching with the 'execute_on' parameter

> What could be going wrong here, or what have I misunderstood?

I think it s all a matter of scheduling of execution. Which is why it's easier to do things on-the-fly and evaluate time derivatives using the proper machinery rather than trying to freeze old values and compute rates manually

**Reply by EdSheltonUKAEA** [reply]
Many thanks again for all your help with this, and developing the Time derivative functor output for me.  Also for your reply on the "bad_cast" error in the other thread.  I've got other work I need to get on with, so I can put further devlopment of this aside for a few days.  If you could  let me know when the time derivative functor is available on moose/master I'll then do a full moose update for my application and try to make use of it from there.  If that fixes my problems in the simplified model by doing what I'm hoping and solves OK,  then I'll apply it to the more complicated full model and resolve the bad_cast error by debugging once I'm confident that should work as planned.

> Have you paste fm_x_now somewhere so I can take a look at what it does and which parameters you passed?
Here are the relevant bits of my code for the FunctorMaterials and AuxKernels.
- Cbar_H_m is the non-linear variable to be solved.
- mu_H, nu_tr1_oc1, Ctild_tr1_oc0 are other FunctorMaterials calculated from Cbar_H_m.
- delta_t is obtained from the TimeStep postprocessor

> [AuxKernels]
>   [Cbar_H_tr1_tot_av]
>     type = FunctorMaterialRealAux
>     variable = elavCbar_H_tr1_tot
>     functor = fmCbar_H_tr1_tot
>   []
>   [Cbar_H_tr1_tot]
>     type = ProjectionAux
>     variable = Cbar_H_tr1_tot
>     v = elavCbar_H_tr1_tot
>   []
> []
> 
> [FunctorMaterials]
>   [Cbar_H_tr1_oc1] 
>     type = ADParsedFunctorMaterial
>     property_name = fmCbar_H_tr1_tot
>     functor_names = 'Cbar_H_m mu_H nu_tr1_oc1 Ctild_tr1_oc0'
>     expression = 'Cbar_H_m * Ctild_tr1_oc0 * mu_H / nu_tr1_oc1'
>     execute_on = TIMESTEP_END
>   []
> 
>   [Cbar_H_tr1_oc1_prev] 
>     type = ADParsedFunctorMaterial
>     property_name = prevfmCbar_H_tr1_tot #_oc1
>     functor_names = 'Cbar_H_m mu_H nu_tr1_oc1 Ctild_tr1_oc0'
>     expression = 'Cbar_H_m * Ctild_tr1_oc0 * mu_H / nu_tr1_oc1'
>     execute_on = TIMESTEP_BEGIN
>   []
> 
>   [H_detrap_rate]
>     type = ADParsedFunctorMaterial
>     property_name = net_detrap_rate_H_tr1_fm
>     functor_names = 'fmCbar_H_tr1_tot prevfmCbar_H_tr1_tot delta_t' 
>     expression = 'if(delta_t>0, (prevfmCbar_H_tr1_tot - fmCbar_H_tr1_tot) / delta_t, 0)'    
>   []
> []
> 
> [Postprocessors]
>   [delta_t]
>     type = TimestepSize
>   []
> []

**Reply by GiudGiud** [reply]
It is available now
https://mooseframework.inl.gov/source/functormaterials/GenericFunctorTimeDerivativeMaterial.html

So the final quantity you are trying to compute here is H_detrap_rate right? the other 2 aux-variables and 2 functor materials are just accessories or outputs ?

Unfortunately, while this first approach is straightforward, it does not work because the time derivative of the parsed functor has not been implemented.  This approach is not free of caveats either.  I ll see if I can make it work end of this month. 

```
[Cbar_H_tr1_oc1]
  type = ADParsedFunctorMaterial
  property_name = fmCbar_H_tr1_tot
  functor_names = 'Cbar_H_m mu_H nu_tr1_oc1 Ctild_tr1_oc0'
  expression = 'Cbar_H_m * Ctild_tr1_oc0 * mu_H / nu_tr1_oc1'
[]
[H_detrap_rate]
    type = ADGenericFunctorTimeDerivativeMaterial
    prop_names = 'net_detrap_rate_H_tr1_fm'
    prop_values = 'fmCbar_H_tr1_tot'
[]
```

But this approach should work
```
[Cbar_H_tr1_oc1]
  type = ADParsedFunctorMaterial
  property_name = net_detrap_rate_H_tr1_fm
  functor_names = 'Cbar_H_m mu_H nu_tr1_oc1 Ctild_tr1_oc0 Cbar_H_m_dt mu_H_dt nu_tr1_oc1_dt Ctild_tr1_oc0_dt'
  expression = 'Cbar_H_m_dt * Ctild_tr1_oc0 * mu_H / nu_tr1_oc1 + Cbar_H_m * Ctild_tr1_oc0_dt * mu_H / nu_tr1_oc1 + Cbar_H_m * Ctild_tr1_oc0 * mu_H_dt / nu_tr1_oc1 - nu_tr1_oc1_dt * Cbar_H_m * Ctild_tr1_oc0 * mu_H / nu_tr1_oc1 / nu_tr1_oc1'
[]
[H_detrap_rate]
    type = ADGenericFunctorTimeDerivativeMaterial
    prop_names = 'Cbar_H_m_dt mu_H_dt nu_tr1_oc1_dt Ctild_tr1_oc0_dt'
    prop_values = 'Cbar_H_m mu_H nu_tr1_oc1 Ctild_tr1_oc0'
[]
```
you are likely the first person trying to solve a complex problem by chaining functors from the input file (rather than coding a new kernel here?). I m happy to keep improving this if it does not work

One of my concerns is that if you are planning to use this as a time derivative term in an equation, we'll need to use an ADTimeKernel derived kernel rather than a simple source term. However the internal TIME vector wont be populated

**Reply by EdSheltonUKAEA** [reply]
Thank you so much for this.  As indicated in my previous message, I probably won't get round to applying this in detail for a few more days.  But I'll be sure to let you know how it goes.
However, for my initial application, the terms mu_H and nu_tr1_oc1 don't change with time so their derivatives will zero and I can simplify the expression to omit these.  But if I develop the use of this further, this time independence won't necessarily apply, so I'll include a comment to remind myself to put these back in.

**Reply by GiudGiud** [reply]
sounds good. thanks for letting me know.
